File: grafuri/Java/javap1/src/p2.java
import java.util.List;
import java.util.Stack;

/**
 * Clasa principala unde veti implementa sortarea topologica.
 * 
 * @author Claudia Cardei
 * 
 */
public class p2 {

	public static Stack<Cell> cellStack = new Stack<Cell>();

	public static void DFS(Cell cell) {
		// TODO: Implementati DFS. La finalul parcurgerii unei celule, adaugati
		// celula in cellStack.
		int t = 0;
		List<Cell> dependent = cell.getDependentCells();
		cell.setColor(Cell.Color.GRAY);
		cell.setInitTime(t);
		++t;
		for(int i = 0; i < dependent.size(); i++) {
			if (dependent.get(i).getColor() == Cell.Color.WHITE) {
				DFS(dependent.get(i));
			}
		}
		// celula in cellStack.
		cell.setColor(Cell.Color.BLACK);
		cell.setFinishTime(t);
		cellStack.add(cell);
	}

	public static void main(String[] args) {
		Reader reader = new Reader("date.in");
		List<Cell> cells = reader.parse();

		for(int i = 0; i < cells.size(); i++) {
			if (cells.get(i).getColor() == Cell.Color.WHITE)
				DFS(cells.get(i));
		}
		
		
		while(!cellStack.isEmpty()) {
			System.out.println(cellStack.pop().toString());
		}

	}
}

File: grafuri/C++/p2/p2.cpp
#include <iostream>
#include <fstream>
#include <map>
#include <sstream>
#include <stack>
#include <string>
#include <vector>

#include "Cell.h"

using namespace std;

stack<Cell*> cell_stack;

vector<Cell*> read_file_input(string file_name) {
    map<string, Cell*> cells;
    vector<Cell*> final_cells;
    
    ifstream in("date.in");
    string formula;
    
    while (getline(in, formula)) {
        stringstream formulastream(formula);
        string token;
        formulastream >> token;
        
        Cell* cell = NULL;
        if (cells.find(token) == cells.end()) {
            cell = new Cell(token);
            cells.insert(make_pair(token, cell));
            
            final_cells.push_back(cell);
        } else {
            cell = cells[token];
        }
        
        while (formulastream >> token) {
            if (token.length() == 2) {
                Cell* dependentCell = new Cell(token);
                if (cells.find(token) == cells.end()) {            
                    cells.insert(make_pair(token, dependentCell));
                    
                    final_cells.push_back(dependentCell);
                }
                  
                cells[token]->add_dependent_cell(cell);
            }
        }
    }
    
    return final_cells;
}

void DFS(Cell* cell) {
    //TODO: Implementati DFS. La finalul parcurgerii unei celule, adaugati
	// celula in cell_stack.
	
}

int main() {
    vector<Cell*> cells = read_file_input("date.in");
    
    //TODO: Afisati parcurgerea celulelor.

    return 0;
}
